#### Building the model with CNN

In [65]:


# Convert target data to one-hot encoded format
train_y_encoded = to_categorical(train_y, num_classes=3)
test_y_encoded = to_categorical(test_y, num_classes=3)

# Create the model
model_CNN = Sequential()
model_CNN.add(Conv2D(32, (3, 3), input_shape=(50, 50,1), activation='relu'))
model_CNN.add(MaxPooling2D(pool_size=(2, 2)))
model_CNN.add(BatchNormalization())
model_CNN.add(Dropout(0.2))
model_CNN.add(Conv2D(32, (3, 3), activation='relu'))
model_CNN.add(MaxPooling2D(pool_size=(2, 2)))
model_CNN.add(BatchNormalization())
model_CNN.add(Dropout(0.2))
model_CNN.add(Flatten())
model_CNN.add(Dense(512, activation='relu'))
model_CNN.add(BatchNormalization())
model_CNN.add(Dropout(0.2))
model_CNN.add(Dense(256, activation='relu'))
model_CNN.add(BatchNormalization())
model_CNN.add(Dropout(0.2))
model_CNN.add(Dense(3, activation='softmax'))  # Change units to 3 for three classes

# Compile the model
model_CNN.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
history = model_CNN.fit(train_x, train_y_encoded, batch_size=64, epochs=10, validation_split=0.1)

# Evaluate the model
test_loss, test_accuracy = model_CNN.evaluate(test_x, test_y_encoded)
print("Test Accuracy: {:.2f}%".format(test_accuracy * 100))


Epoch 1/10
488/488 [==============================] - 134s 266ms/step - loss: 1.0104 - accuracy: 0.5341 - val_loss: 1.3690 - val_accuracy: 0.4035
Epoch 2/10
488/488 [==============================] - 131s 268ms/step - loss: 0.8585 - accuracy: 0.6030 - val_loss: 0.8094 - val_accuracy: 0.6279
Epoch 3/10
488/488 [==============================] - 137s 281ms/step - loss: 0.7904 - accuracy: 0.6405 - val_loss: 0.7398 - val_accuracy: 0.6608
Epoch 4/10
488/488 [==============================] - 134s 275ms/step - loss: 0.7495 - accuracy: 0.6581 - val_loss: 0.7488 - val_accuracy: 0.6571
Epoch 5/10
488/488 [==============================] - 139s 284ms/step - loss: 0.7171 - accuracy: 0.6776 - val_loss: 0.7708 - val_accuracy: 0.6449
Epoch 6/10
488/488 [==============================] - 137s 281ms/step - loss: 0.6962 - accuracy: 0.6904 - val_loss: 0.6775 - val_accuracy: 0.6896
Epoch 7/10
488/488 [==============================] - 131s 268ms/step - loss: 0.6748 - accuracy: 0.6987 - val_loss: 0.7875 -

In [66]:
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions
predictions = model_CNN.predict(test_x)
predicted_classes = np.argmax(predictions, axis=1)

# Convert one-hot encoded labels back to original form
true_classes = np.argmax(test_y_encoded, axis=1)

# Build evaluation matrix
evaluation_matrix = confusion_matrix(true_classes, predicted_classes)
classification_report = classification_report(true_classes, predicted_classes)

# Print evaluation matrix and classification report
print("Evaluation Matrix:")
print(evaluation_matrix)
print("\nClassification Report:")
print(classification_report)


136/136 [==============================] - 3s 19ms/step
Evaluation Matrix:
[[ 983  172  136]
 [ 258 1010  269]
 [ 204  200 1102]]

Classification Report:
              precision    recall  f1-score   support

           0       0.68      0.76      0.72      1291
           1       0.73      0.66      0.69      1537
           2       0.73      0.73      0.73      1506

    accuracy                           0.71      4334
   macro avg       0.71      0.72      0.71      4334
weighted avg       0.72      0.71      0.71      4334



In [67]:
# Compute AUC for CNN model
auc_CNN = roc_auc_score(test_y_encoded, model_CNN.predict(test_x), multi_class='ovr')
print("AUC (CNN Model): {:.4f}".format(auc_CNN))

136/136 [==============================] - 3s 18ms/step
AUC (CNN Model): 0.8794


With an AUC of 0.8794, the CNN model demonstrates a strong discriminatory power. Generally, an AUC value above 0.8 is considered good, suggesting that the model has excellent performance in distinguishing between the classes. The AUC value of 0.8794 indicates that the CNN model is performing well in terms of classification accuracy.In summary, based on the provided AUC value of 0.8794, the CNN model's performance can be considered good. 

Based on the provided evaluation matrix and classification report, the model's performance can be considered good.The accuracy of the model is 0.71, indicating that it correctly predicts the class for approximately 71% of the instances. The precision, recall, and F1-score for each class are also reasonably high, ranging from 0.68 to 0.73. These metrics suggest that the model performs well in terms of correctly identifying instances of each class.

The macro average F1-score of 0.71 and the weighted average F1-score of 0.71 further support the conclusion that the model's performance is good overall. The macro average considers the equal importance of each class, while the weighted average accounts for the class distribution.In summary, based on the provided evaluation metrics and classification report, the model's performance appears to be good, with reasonably high accuracy, precision, recall, and F1-scores.

### Now doing the combined dataset with DNN

In [78]:
# Convert target data to one-hot encoded format
train_y_encoded = to_categorical(train_y, num_classes=3)
test_y_encoded = to_categorical(test_y, num_classes=3)

# Create the model
model_DNN = Sequential()
model_DNN.add(Flatten(input_shape=(50, 50, 1)))
model_DNN.add(Dense(32, activation='relu'))
model_DNN.add(BatchNormalization())
model_DNN.add(Dropout(0.2))
model_DNN.add(Dense(32, activation='relu'))
model_DNN.add(BatchNormalization())
model_DNN.add(Dropout(0.2))
model_DNN.add(Dense(512, activation='relu'))
model_DNN.add(BatchNormalization())
model_DNN.add(Dropout(0.2))
model_DNN.add(Dense(256, activation='relu'))
model_DNN.add(BatchNormalization())
model_DNN.add(Dropout(0.2))
model_DNN.add(Dense(3, activation='softmax'))  # Change units to 3 for three classes

# Compile the model
model_DNN.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
history = model_DNN.fit(train_x, train_y_encoded, batch_size=64, epochs=10, validation_split=0.1)

# Evaluate the model
test_loss, test_accuracy = model_DNN.evaluate(test_x, test_y_encoded)
print("Test Accuracy: {:.2f}%".format(test_accuracy * 100))


Epoch 1/10
488/488 [==============================] - 12s 13ms/step - loss: 1.1505 - accuracy: 0.4404 - val_loss: 0.9925 - val_accuracy: 0.5120
Epoch 2/10
488/488 [==============================] - 6s 13ms/step - loss: 0.9897 - accuracy: 0.5195 - val_loss: 0.9333 - val_accuracy: 0.5521
Epoch 3/10
488/488 [==============================] - 6s 13ms/step - loss: 0.9527 - accuracy: 0.5421 - val_loss: 0.8913 - val_accuracy: 0.5757
Epoch 4/10
488/488 [==============================] - 5s 9ms/step - loss: 0.9370 - accuracy: 0.5464 - val_loss: 0.8818 - val_accuracy: 0.5893
Epoch 5/10
488/488 [==============================] - 5s 10ms/step - loss: 0.9270 - accuracy: 0.5525 - val_loss: 0.8750 - val_accuracy: 0.5873
Epoch 6/10
488/488 [==============================] - 6s 12ms/step - loss: 0.9171 - accuracy: 0.5619 - val_loss: 0.8673 - val_accuracy: 0.5924
Epoch 7/10
488/488 [==============================] - 5s 10ms/step - loss: 0.9090 - accuracy: 0.5668 - val_loss: 0.8942 - val_accuracy: 0.5673

In [79]:
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix

# Make predictions
predictions = model_DNN.predict(test_x)
predicted_classes = np.argmax(predictions, axis=1)

# Convert one-hot encoded labels back to original form
true_classes = np.argmax(test_y_encoded, axis=1)

# Build evaluation matrix
evaluation_matrix = confusion_matrix(true_classes, predicted_classes)
classification_report = classification_report(true_classes, predicted_classes)

# Print evaluation matrix and classification report
print("Evaluation Matrix:")
print(evaluation_matrix)
print("\nClassification Report:")
print(classification_report)


136/136 [==============================] - 1s 3ms/step
Evaluation Matrix:
[[ 826  367   98]
 [ 363 1014  160]
 [ 304  570  632]]

Classification Report:
              precision    recall  f1-score   support

           0       0.55      0.64      0.59      1291
           1       0.52      0.66      0.58      1537
           2       0.71      0.42      0.53      1506

    accuracy                           0.57      4334
   macro avg       0.59      0.57      0.57      4334
weighted avg       0.60      0.57      0.57      4334



In [80]:
# Compute AUC for DNN model
auc_DNN = roc_auc_score(test_y_encoded, model_DNN.predict(test_x), multi_class='ovr')
print("AUC (DNN Model): {:.4f}".format(auc_DNN))

136/136 [==============================] - 0s 3ms/step
AUC (DNN Model): 0.7754



Based on the provided information, the DNN (Deep Neural Network) model achieves an AUC (Area Under the Curve) value of 0.7754.

The AUC is a commonly used metric to evaluate the performance of binary classification models. It represents the model's ability to distinguish between positive and negative instances. Generally, a higher AUC value indicates better model performance.

With an AUC of 0.7754, the DNN model demonstrates a moderate discriminatory power. While it is not as high as an AUC above 0.8, it still suggests that the model has some ability to discriminate between the classes.

In summary, based on the provided AUC value of 0.7754, the DNN model's performance can be considered moderate.

Based on the provided evaluation matrix and classification report, the model's performance can be considered fair to moderate.The accuracy of the model is 0.57, indicating that it correctly predicts the class for approximately 57% of the instances. The precision, recall, and F1-score for each class range from 0.52 to 0.71, with varying performance across the classes.The macro average F1-score of 0.57 and the weighted average F1-score of 0.57 suggest that the model's overall performance is fair to moderate. The macro average considers the equal importance of each class, while the weighted average accounts for the class distribution.

In summary, based on the provided evaluation metrics and classification report, the model's performance can be considered fair to moderate. 